singer/src/main/java/com/pinterest/singer/writer/KafkaProducerMetricsMonitor.java

```diff
@@ -0,0 +1,86 @@
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.pinterest.singer.writer;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.pinterest.singer.metrics.OpenTsdbMetricConverter;
+import com.pinterest.singer.thrift.configuration.KafkaProducerConfig;
+
+/**
+ * Responsible for pulling metrics from {@link KafkaProducer} and copying them
+ * to Ostrich so they can be accessed and forwarded. This helps provide
+ * additional instrumentation on Singer and how it's performing.
+ */
+public class KafkaProducerMetricsMonitor implements Runnable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaProducerMetricsMonitor.class);
+  public static final Set<String> PRODUCER_METRICS_WHITELIST = new HashSet<>(
+      Arrays.asList("buffer-total-bytes", "buffer-available-bytes"));
+  // sample every 60seconds
+  private static final int SAMPLING_INTERVAL = 60_000;
+
+  @Override
+  public void run() {
+    while (true) {
+      try {
+        publishKafkaProducerMetricsToOstrich();
+      } catch (Exception e) {
+        LOG.warn("Error publishing KafkaProducer metrics", e);
+      }
+      try {
+        Thread.sleep(SAMPLING_INTERVAL);
+      } catch (InterruptedException e) {
+        LOG.warn("KafkaProducerMetricsMonitor thread interrupted, exiting");
+        break;
+      }
+    }
+  }
+
+  @SuppressWarnings({ "deprecation" })
+  protected void publishKafkaProducerMetricsToOstrich() {
+    Map<KafkaProducerConfig, KafkaProducer<byte[], byte[]>> producers = KafkaProducerManager
+        .getInstance().getProducers();
+    for (Entry<KafkaProducerConfig, KafkaProducer<byte[], byte[]>> kafkaProducerEntry : producers
+        .entrySet()) {
+      KafkaProducerConfig key = kafkaProducerEntry.getKey();
+      String signature = convertSignatureToTag(key);
+      Map<MetricName, ? extends Metric> metrics = kafkaProducerEntry.getValue().metrics();
+      for (Entry<MetricName, ? extends Metric> entry : metrics.entrySet()) {
+        if (PRODUCER_METRICS_WHITELIST.contains(entry.getKey().name())) {
+          OpenTsdbMetricConverter.gauge(entry.getKey().name(), entry.getValue().value(),
+              "cluster=" + signature);
+        }
+      }
+    }
+  }
+
+  public static String convertSignatureToTag(KafkaProducerConfig key) {
+    return key.getKafkaClusterSignature().replace("/discovery/", "").replace("/prod", "")
+        .replace("/", "_");
```

Comment:
Do we need to account for `tls` serversets in this conversion?

Reply:
My current thought is to have other names be appended to the setup since the key is different so if it's any extension it really would represent a different cluster signature meaning a different producer.